kubernetes/structures_replication_controller.go

```diff
@@ -32,16 +55,52 @@ func expandReplicationControllerSpec(rc []interface{}) (v1.ReplicationController
 	obj.MinReadySeconds = int32(in["min_ready_seconds"].(int))
 	obj.Replicas = ptrToInt32(int32(in["replicas"].(int)))
 	obj.Selector = expandStringMap(in["selector"].(map[string]interface{}))
-	podSpec, err := expandPodSpec(in["template"].([]interface{}))
+
+	template, err := expandReplicationControllerTemplate(in["template"].([]interface{}), obj.Selector)
 	if err != nil {
 		return obj, err
 	}
-	obj.Template = &v1.PodTemplateSpec{
-		ObjectMeta: metav1.ObjectMeta{
-			Labels: obj.Selector,
-		},
-		Spec: podSpec,
+
+	obj.Template = &template
+
+	return obj, nil
+}
+
+func expandReplicationControllerTemplate(rct []interface{}, selector map[string]string) (v1.PodTemplateSpec, error) {
+	obj := v1.PodTemplateSpec{}
+	in := rct[0].(map[string]interface{})
+
+	// Get user defined metadata
+	metadata := expandMetadata(in["metadata"].([]interface{}))
+
+	// Add or merge labels from selector to ensure proper selection of pods by the replication controller
+	if metadata.Labels == nil {
```

Comment:
I'd like to understand more about why you chose this approach of adding the selector labels to metadata.
I understand what you're trying to accomplish, what I'm trying to understand still is why?

Reply:
The opposite way would indeed feel more natural.
However, if it was only up to me, I'd remove the *automagic* behavior altogether and leave it up to the user to set the appropriate labels and selector.
Kubernetes is a fast moving target, doing the most basic mapping for all resources is already a big task, so we should probably stick to it, and not try to hide such properties.

Reply:
I'm totally with you on that.
I think we should get rid of the "magic".

Reply:
Done.